src/test/java/com/commercetools/sync/producttypes/helpers/ProductTypeReferenceResolverTest.java

```diff
@@ -0,0 +1,295 @@
+            ProductTypeDraftBuilder.of(productTypeDraft)
+                                   .attributes(emptyList())
+                                   .build();
+
+        // test and assertion
+        assertThat(productTypeReferenceResolver.resolveReferences(productTypeDraft))
+            .isCompletedWithValue(expectedResolvedProductTypeDraft);
+    }
+
+    @Test
+    void resolveReferences_WithNoNestedTypeReferences_ShouldResolveReferences() {
+        // preparation
+
+        final AttributeDefinitionDraft attributeDefinitionDraft = AttributeDefinitionDraftBuilder
+            .of(StringAttributeType.of(), "string attr", ofEnglish("string attr label"), true)
+            .build();
+
+        final ProductTypeDraft productTypeDraft =
+            ProductTypeDraftBuilder.of("foo", "foo", "desc", singletonList(attributeDefinitionDraft))
+                                   .build();
+
+        final ProductTypeSyncOptions syncOptions =
+            ProductTypeSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+
+        final ProductTypeService productTypeService = mock(ProductTypeService.class);
+        when(productTypeService.fetchCachedProductTypeId(any()))
+            .thenReturn(CompletableFuture.completedFuture(Optional.of("foo")));
+
+        final ProductTypeReferenceResolver productTypeReferenceResolver = new ProductTypeReferenceResolver(syncOptions,
+            productTypeService);
+
+
+        final ProductTypeDraftDsl expectedResolvedProductTypeDraft =
+            ProductTypeDraftBuilder.of(productTypeDraft)
+                                   .attributes(singletonList(attributeDefinitionDraft))
+                                   .build();
+
+        // test and assertion
+        assertThat(productTypeReferenceResolver.resolveReferences(productTypeDraft))
+            .isCompletedWithValue(expectedResolvedProductTypeDraft);
+    }
+
+    @Test
+    void resolveReferences_WithOneNestedTypeWithExistingProductTypeReference_ShouldResolveReferences() {
+        // preparation
+        final NestedAttributeType nestedAttributeType = NestedAttributeType.of(ProductType.reference("x"));
+        final AttributeDefinitionDraft attributeDefinitionDraft =
+            AttributeDefinitionDraftBuilder.of(nestedAttributeType, "foo", ofEnglish("foo"), true)
+                                           .build();
+
+        final ProductTypeDraft productTypeDraft =
+            ProductTypeDraftBuilder.of("foo", "foo", "desc",
+                singletonList(attributeDefinitionDraft))
+                                   .build();
+
+        final ProductTypeSyncOptions syncOptions =
+            ProductTypeSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+
+        final ProductTypeService productTypeService = mock(ProductTypeService.class);
+        when(productTypeService.fetchCachedProductTypeId(any()))
+            .thenReturn(CompletableFuture.completedFuture(Optional.of("foo")));
```

Comment:
in mocks for any id, you are returning `foo` maybe return `x` like the prepared nested attribute?

Reply:
but in the prepared nested attribute the `"x"` is a key actually. While this `"foo"` is an id.